flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/FixedLengthRecordSorter.java

```diff
@@ -447,11 +447,13 @@ public void writeToOutput(final ChannelWriterOutputView output, final int start,
 				num -= recordsPerSegment;
 			} else {
 				// partially filled segment
-				for (; num > 0; num--) {
+				for (; num > 0 && offset <= this.lastEntryOffset; num--, offset += this.recordSize) {
```

Comment:
Yes, it is. It only support to write data inside a `MemorySegment` previously, it's work well before because it only called with `num = 1`. I would create a separate JIRA for this.